jmx-scraper/src/main/java/io/opentelemetry/contrib/jmxscraper/client/JmxRemoteClient.java

```diff
@@ -0,0 +1,149 @@
+import javax.security.sasl.RealmCallback;
+
+public class JmxRemoteClient {
+
+  private static final Logger logger = Logger.getLogger(JmxRemoteClient.class.getName());
+
+  private final JMXServiceURL url;
+  @Nullable private String userName;
+  @Nullable private String password;
+  @Nullable private String profile;
+  @Nullable private String realm;
+  private boolean sslRegistry;
+
+  private JmxRemoteClient(JMXServiceURL url) {
+    this.url = url;
+  }
+
+  public static JmxRemoteClient createNew(String host, int port) {
+    return new JmxRemoteClient(buildUrl(host, port));
+  }
+
+  public static JmxRemoteClient createNew(String url) {
+    return new JmxRemoteClient(buildUrl(url));
+  }
+
+  @CanIgnoreReturnValue
+  public JmxRemoteClient userCredentials(String userName, String password) {
+    this.userName = userName;
+    this.password = password;
+    return this;
+  }
+
+  @CanIgnoreReturnValue
+  public JmxRemoteClient withRemoteProfile(String profile) {
+    this.profile = profile;
+    return this;
+  }
+
+  @CanIgnoreReturnValue
+  public JmxRemoteClient withRealm(String realm) {
+    this.realm = realm;
+    return this;
+  }
+
+  @CanIgnoreReturnValue
+  public JmxRemoteClient withSslRegistry() {
+    this.sslRegistry = true;
+    return this;
+  }
+
+  public JMXConnector connect() throws IOException {
+    Map<String, Object> env = new HashMap<>();
+    if (userName != null && password != null) {
+      env.put(JMXConnector.CREDENTIALS, new String[] {userName, password});
+    }
+
+    if (profile != null) {
+      env.put("jmx.remote.profile", profile);
+    }
+
+    try {
+      // Not all supported versions of Java contain this Provider
+      // Also it might not be accessible due to java.security.sasl module not accessible
+      Class<?> klass = Class.forName("com.sun.security.sasl.Provider");
+      Provider provider = (Provider) klass.getDeclaredConstructor().newInstance();
+      Security.addProvider(provider);
+
+      env.put(
+          "jmx.remote.sasl.callback.handler",
+          (CallbackHandler)
+              callbacks -> {
+                for (Callback callback : callbacks) {
+                  if (callback instanceof NameCallback) {
+                    ((NameCallback) callback).setName(userName);
+                  } else if (callback instanceof PasswordCallback) {
+                    char[] pwd = password == null ? null : password.toCharArray();
+                    ((PasswordCallback) callback).setPassword(pwd);
+                  } else if (callback instanceof RealmCallback) {
+                    ((RealmCallback) callback).setText(realm);
+                  } else {
+                    throw new UnsupportedCallbackException(callback);
+                  }
+                }
+              });
+    } catch (ReflectiveOperationException e) {
+      logger.log(Level.WARNING, "SASL unsupported in current environment: " + e.getMessage());
+    }
```

Comment:
Factor out the top half of this method to a new method called `buildEnv()` to improve readability/length.